internal/compiler/generator/cpp.rs

```diff
@@ -82,17 +85,96 @@ mod cpp_ast {
     ///A full C++ file
     #[derive(Default, Debug)]
     pub struct File {
+        pub is_cpp_file: bool,
         pub includes: Vec<String>,
         pub after_includes: String,
         pub namespace: Option<String>,
         pub declarations: Vec<Declaration>,
+        pub resources: Vec<Declaration>, // For now inline, maybe #embed in the future
         pub definitions: Vec<Declaration>,
     }
 
+    impl File {
+        pub fn split_off_cpp_files(&mut self, header_file_name: String, count: usize) -> Vec<File> {
+            let mut cpp_files = Vec::with_capacity(count);
+            if count == 0 {
+                return cpp_files;
+            }
+
+            let mut definitions = Vec::new();
+
+            let mut i = 0;
+            while i < self.definitions.len() {
+                if matches!(&self.definitions[i], Declaration::Function(fun) if fun.template_parameters.is_some())
+                {
+                    i += 1;
+                    continue;
+                }
+                definitions.push(self.definitions.remove(i));
+            }
+
+            // Any definition in the header file is inline.
+            self.definitions.iter_mut().for_each(|def| match def {
+                Declaration::Function(f) => f.is_inline = true,
+                Declaration::Var(v) => v.is_inline = true,
+                _ => {}
```

Comment:
what if count == 0?  we returned earlier.

Reply:
Good catch - thanks.